src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,146 @@
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> t = tail;
+        Node<T> newNode = new Node<>(t, value, null);
+        tail = newNode;
+        if (t == null) {
+            head = newNode;
+        } else {
+            t.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> var = node(index).prev;
+            Node<T> newNode = new Node<>(var, value, node(index));
+            node(index).prev = newNode;
+            if (var == null) {
+                head = newNode;
+            } else {
+                var.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return node(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        final T element = node(index).value;
+        Node<T> var = node(index);
+        var.value = value;
+        return element;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        final T element = node(index).value;
+        Node<T> prev = node(index).prev;
+        Node<T> next = node(index).next;
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+        }
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+        }
+        size--;
+        return element;
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (int i = 0; i < size; i++) {
+                if (node(i).value == null) {
+                    remove(i);
+                    return true;
+                }
+            }
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (object.equals(node(i).value)) {
+                    remove(i);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private Node<T> node(int index) {
```

Comment:
```suggestion
    private Node<T> getNode(int index) {
```
зазвичай в неймінгу методів використовуються дієслова